File: Teacher/05Richer/step1_src_完成1/mapimp.cpp
///======================================================================
///  Project:   Richer02
/// FileName:	mapimp.cpp
///     Desc:   Richer 01
///   Author:	Chen Wei
///======================================================================
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
#include "mapimp.h"
#include "block.h"
#include "player.h"
MapImp::MapImp()
{
    //ctor
}

MapImp::~MapImp()
{
    //dtor
    for(Block* block : blocks) {
        delete block;
    }
}

void MapImp::addBlock(Block* block)
{
    blocks.push_back(block);
}
Block* MapImp::getBlock(int index)
{
    return blocks[index];
}
void MapImp::show() const
{
    //OutDevice & dev = OutDevice::getConsole();

    vector<Block*> temps(blocks);
    sort(temps.begin(), temps.end(), [](Block * b1, Block * b2) {
        if(b1->getRow() == b2->getRow()) {
            return b1->getCol() < b2->getCol();
        } else {
            return b1->getRow() < b2->getRow();
        }
    });

    Block* b1 = Players::getPlayers().getFirst()->getStand();
    Block* b2 = Players::getPlayers().getLast()->getStand();

    int row = 0;
    int col = 0;
    for(Block* block : temps) {
        int r = block->getRow();
        int c = block->getCol();
        for(int i = row; i < r; ++i) {
            cout << endl;
            col = 0;
        }
        for(int i = col; i < c; ++i) {
            cout << "   ";
        }
        if(block == b1) {
            cout << "["<<block->name() << "]";
        } else if(block == b2) {
            cout << "("<<block->name() << ")";
        } else {
            cout <<" "<<block->name() << " ";
        }
        row = r;
        col = c + 1;
    }
    cout << endl;
}

File: Mine/05Monopoly/Monopoly-Q2特别版/src/map/MapImp.java
package map;
import java.util.*;
import block.Block;
import block.WallBlock;
import game.Game;
import player.Player;

public class MapImp {
	protected Vector<Block> blocks;
	protected int row;
	protected int col;

	MapImp(int r, int c) {
		row = r;
		col = c;
		blocks = new Vector<Block>();
	}

	public void show() {
		Block[][] blockList = new Block[row][col];

		Block block = new WallBlock();
		for (int i = 0; i < blockList.length; i++){
			for (int j = 0; j < blockList[i].length; j++) {
				blockList[i][j] = block;
			}
		}

		for (Block specificBlock : blocks) {
			blockList[specificBlock.getRow()][specificBlock.getCol()] = specificBlock;
		}
		
		Queue<Player> players = Game.getGame().players;
		for (int i = 0; i < blockList.length; i++) {
			System.out.println();
			System.out.println();
            for (int j = 0; j < blockList[i].length; j++){
            	boolean flag = false;
            	for (Player player:players) {
            		if ((player.x == i) && (player.y == j)) {
            			System.out.print(player.show());
            			flag = true;
            		}
            	}
            	
            	if(!flag) {
            		System.out.print(" "+blockList[i][j].toString()+" ");
            	}
            }
        }
	}

	public void addBlock(Block block) {
		blocks.add(block);
	}

	public Block getBlock(int index){
		return blocks.get(index);
	}

	public void setBlockVector(Vector<Block> blockVector) {
		blocks = blockVector;
	}
	
	public Vector<Block> getblockVector() {
		return blocks;
	}

	public int getRow() {
		return row;
	}

	public int getCol() {
		return col;
	}
}